lucene/core/src/java/org/apache/lucene/util/automaton/IntSet.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.util.automaton;
+
+import java.util.Arrays;
+
+public abstract class IntSet {
+    int[] values;
+
+    int hashCode;
```

Comment:
I think the hashCode should be private to the abstract class and anything else should call hashCode() (including equals method). Then subclasses that are mutable could return true hashCode of their content (cached or computed).

Reply:
In addition, fields that are initialized in constructors and never changed should always be declared final.